blockchain/client.go

```diff
@@ -77,10 +78,105 @@ func (c *Client) StartEventsListening() (func(), <-chan error, error) {
 			case <-done:
 				return
 			case set := <-sub.Chan():
+				c.processSystemEventsStorageChanges(
+					set.Changes,
+					meta,
+					key,
+					set.Block,
+				)
+			}
+		}
+	}()
+
+	once := sync.Once{}
+	c.cancelListening = func() {
+		once.Do(func() {
+			done <- struct{}{}
+			sub.Unsubscribe()
+			c.isListening = 0
+		})
+	}
+
+	return c.cancelListening, c.errsListening, nil
+}
+
+// RegisterEventsListener subscribes given callback to blockchain events. There is a begin parameter which
+// can be used to get events from blocks older than the latest block. If begin is greater than the latest
+// block number, the listener will start from the latest block.
+func (c *Client) RegisterEventsListener(begin types.BlockNumber, callback EventsListener) (context.CancelFunc, error) {
+	var idx int
+	for i := 0; i <= math.MaxInt; i++ {
```

Comment:
this for loop looks strange. how it works?

Reply:
It finds the smallest available key of the map.

Reply:
why do we need map there? can't we use an array?

Reply:
at the moment it's a map where key is an index...

Reply:
We can use an array or a slice here, but it will be harder to deal with events listeners deletion. Using array we should define a max number of expected subscribers or deal with its grows and shrinkage somehow. Using slice we should track deletions and update indexes of shifted listeners for their deletion functions. Both options look more complicated to me than the map plus this index assigning function.

Reply:
i see. can't we use `map[listener]bool`? 

Then you don't need to manage indexes. If no, then feel free to keep as is.

Reply:
Yes, cool, we can use function pointers as keys. Updated to `map[*listener]struct{}` and the strange loop is removed.